### ISSUE TITLE
Run distributed training with MPI get blocked

### ISSUE THREAD

**Issue by Applepleple** [facebookarchive/caffe2#1458]
Hi team,

I have been working on using MPI to run distributed training with demo resnet50_trainer.py

When I run the demo with the following command, the demo worked.
mpirun -n 2 --host 10.134.142.142,10.134.142.105 python /home/lwang/caffe2/caffe2/python/examples/resnet50_trainer.py --train_data /home/lwang/dataset/mnist/mnist_train_lmdb/ --num_gpus 2 --num_epochs 1 --epoch_size 6400

When I use a hostfile for openmpi to do this again, I get blocked.

**Command:**
mpirun -n 2 -hostfile /home/hadoop/tmp/mpiexecs/appattempt_1510379431381_0013_000001/.hostfile python /home/hadoop/tmp/mpiexecs/appattempt_1510379431381_0013_000001/MPICH2-resnet50_trainer.py --train_data /home/lwang/dataset/mnist/mnist_train_lmdb/ --num_epochs 1 --epoch_size 6400 --file_store_path /home/lwang/dataset/mnist  --num_gpus 2

**Hostfile:**
10.134.142.105 slots=1
10.134.142.142 slots=1

**Log output:**
Ignoring @/caffe2/caffe2/contrib/nccl:nccl_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops_gpu as it is not a valid file.
Ignoring @/caffe2/caffe2/distributed:file_store_handler_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/distributed:redis_store_handler_ops as it is not a valid file.
INFO:resnet50_trainer:Rank 0 running on GPUs: [0, 1]
INFO:resnet50_trainer:Using epoch size: 6400
INFO:data_parallel_model:Parallelizing model for devices: [0, 1]
INFO:data_parallel_model:Create input and model training operators
INFO:data_parallel_model:Model for GPU : 0
INFO:data_parallel_model:Model for GPU : 1
INFO:data_parallel_model:Adding gradient operators
INFO:data_parallel_model:Add gradient all-reduces for SyncSGD
WARNING:data_parallel_model:Distributed broadcast of computed params is not implemented yet
INFO:data_parallel_model:Post-iteration operators for updating params
INFO:data_parallel_model:Calling optimizer builder function
INFO:data_parallel_model:Add initial parameter sync
WARNING:data_parallel_model:------- DEPRECATED API, please use data_parallel_model.OptimizeGradientMemory() ----- 
WARNING:memonger:NOTE: Executing memonger to optimize gradient memory
INFO:memonger:Memonger memory optimization took 0.0240709781647 secs
WARNING:memonger:NOTE: Executing memonger to optimize gradient memory
INFO:memonger:Memonger memory optimization took 0.0234560966492 secs
Ignoring @/caffe2/caffe2/contrib/nccl:nccl_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops_gpu as it is not a valid file.
Ignoring @/caffe2/caffe2/distributed:file_store_handler_ops as it is not a valid file.
Ignoring @/caffe2/caffe2/distributed:redis_store_handler_ops as it is not a valid file.
INFO:resnet50_trainer:Rank 1 running on GPUs: [0, 1]
INFO:resnet50_trainer:Using epoch size: 6400
INFO:data_parallel_model:Parallelizing model for devices: [0, 1]
INFO:data_parallel_model:Create input and model training operators
INFO:data_parallel_model:Model for GPU : 0
INFO:data_parallel_model:Model for GPU : 1
INFO:data_parallel_model:Adding gradient operators
INFO:data_parallel_model:Add gradient all-reduces for SyncSGD
WARNING:data_parallel_model:Distributed broadcast of computed params is not implemented yet
INFO:data_parallel_model:Post-iteration operators for updating params
INFO:data_parallel_model:Calling optimizer builder function
INFO:data_parallel_model:Add initial parameter sync
WARNING:data_parallel_model:------- DEPRECATED API, please use data_parallel_model.OptimizeGradientMemory() ----- 
WARNING:memonger:NOTE: Executing memonger to optimize gradient memory
INFO:memonger:Memonger memory optimization took 0.148816108704 secs
WARNING:memonger:NOTE: Executing memonger to optimize gradient memory
INFO:memonger:Memonger memory optimization took 0.150115013123 secs

**When I press Ctrl-C to stop it**
^C*** Aborted at 1510470426 (unix time) try "date -d @1510470426" if you are using GNU date ***
PC: @     0x7f3cb215f70d __poll
*** SIGTERM (@0x3f400004fc2) received by PID 20421 (TID 0x7f3cb2844700) from PID 20418; stack trace: ***
    @     0x7f3cb243f390 (unknown)
    @     0x7f3cb215f70d __poll
    @     0x7f3c740e4e58 (unknown)
    @     0x7f3c740db6fb opal_libevent2021_event_base_loop
    @     0x7f3c740a5238 opal_progress
    @     0x7f3c91f2bf65 ompi_request_default_wait_all
    @     0x7e3c49e8f426 ompi_coll_tuned_allreduce_intra_recursivedoubling
    @     0x7f3c91f1c4ce ompi_comm_nextcid
    @     0x7f3c91f18668 ompi_comm_dup_with_info
    @     0x7f3c91f40fbd MPI_Comm_dup
    @     0x7f3cad9549e0 gloo::mpi::Context::Context()
    @     0x7f3cad954d3f gloo::mpi::Context::createManaged()
    @     0x7f3c9778864e caffe2::gloo::CreateCommonWorld<>::RunOnDevice()
    @     0x7f3c9777791b caffe2::Operator<>::Run()
    @     0x7f3c925728e9 caffe2::SimpleNet::DoRunAsync()
    @     0x7f3c9250e7fb caffe2::NetBase::RunAsync()
    @     0x7f3c925034f7 caffe2::Workspace::RunNetOnce()
    @     0x7f3cad8fd29a _ZZN6caffe26python16addGlobalMethodsERN8pybind116moduleEENKUlRKNS1_5bytesEE28_clES6_.isra.2312.constprop.2369
    @     0x7f3cad8fd535 _ZZN8pybind1112cpp_function10initializeIZN6caffe26python16addGlobalMethodsERNS_6moduleEEUlRKNS_5bytesEE28_bJS8_EJNS_4nameENS_5scopeENS_7siblingEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlRNS_6detail13function_callEE1_4_FUNESQ_
    @     0x7f3cad921b03 pybind11::cpp_function::dispatcher()
    @           0x4cb45e PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4c2509 PyEval_EvalCode
    @           0x4f1def (unknown)
*** Aborted at 1510470426 (unix time) try "date -d @1510470426" if you are using GNU date ***
PC: @     0x7fc69f90970d __poll
*** SIGTERM (@0x3ed00005d9a) received by PID 23963 (TID 0x7fc69ffef700) from PID 23962; stack trace: ***
    @     0x7fc69fbe9390 (unknown)
    @     0x7fc69f90970d __poll
    @     0x7fc60f1c4e58 (unknown)
    @     0x7fc60f1bb6fb opal_libevent2021_event_base_loop
    @     0x7fc60f185238 opal_progress
    @     0x7fc62ee5af65 ompi_request_default_wait_all
    @     0x7fc6660f5426 ompi_coll_tuned_allreduce_intra_recursivedoubling
    @     0x7fc62ee4b4ce ompi_comm_nextcid
    @     0x7fc62ee47668 ompi_comm_dup_with_info
    @     0x7fc62ee6ffbd MPI_Comm_dup
    @     0x7fc637a6a6a0 gloo::mpi::Context::Context()
    @     0x7fc637a6a9ff gloo::mpi::Context::createManaged()
    @     0x7fc6352a212e caffe2::gloo::CreateCommonWorld<>::RunOnDevice()
    @     0x7fc6352913bb caffe2::Operator<>::Run()
    @     0x7fc62f6a5a09 caffe2::SimpleNet::DoRunAsync()
    @     0x7fc62f6749ab caffe2::NetBase::RunAsync()
    @     0x7fc62f69bce7 caffe2::Workspace::RunNetOnce()
    @     0x7fc637a1328a _ZZN6caffe26python16addGlobalMethodsERN8pybind116moduleEENKUlRKNS1_5bytesEE28_clES6_.isra.2310.constprop.2367
    @     0x7fc637a13525 _ZZN8pybind1112cpp_function10initializeIZN6caffe26python16addGlobalMethodsERNS_6moduleEEUlRKNS_5bytesEE28_bJS8_EJNS_4nameENS_5scopeENS_7siblingEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlRNS_6detail13function_callEE1_4_FUNESQ_
    @     0x7fc637a37853 pybind11::cpp_function::dispatcher()
    @           0x4cb45e PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4ca8d1 PyEval_EvalFrameEx
    @           0x4c2765 PyEval_EvalCodeEx
    @           0x4c2509 PyEval_EvalCode
    @           0x4f1def (unknown)

OS type is Ubuntu 16.04 LTS;
OpenMPI version is 1.10.2;

**Comment by pietern** [comment]
Thanks for posting the issue, @Applepleple. OpenMPI is a black box to me and I can't give you any hints why it would or would not work with a hostfile. I think OpenMPI uses a different initialization path in the latter case and your stack on abort confirms that the program is hanging in OpenMPI.

I don't think we have folks on our team who would know how to fix this, so I'm closing the issue. I hope you can launch distributed runs through mpirun without hostfile, or through some different means that doesn't require MPI at all (this is what I prefer personally TBH).
